src/Oscoin/Crypto/Blockchain/Eval.hs

```diff
@@ -159,32 +156,27 @@ buildGenesis eval tick txs s =
         Crypto.zeroHash
         s
 
--- | Try to evaluate a block, given an initial state and evaluator.
+-- | Evaluate the transactions contained in the block against the given
+-- state and return the new state and all the produced Receipts.
 evalBlock
-    :: Evaluator st tx o
+    :: forall c st tx o s.
+       (Crypto.Hashable c tx)
+    => Evaluator st tx o
     -> st
     -> Block c tx s
-    -> Either EvalError st
-evalBlock eval st b =
-    foldlM step st (blockData b)
-  where
-    step s tx = map snd (eval tx s)
-
--- | Evaluate all transactions of a 'Blockchain' and return the result.
-evalBlockchain
-    :: Evaluator st tx o
-    -> st
-    -> Blockchain c tx s
-    -> ([(tx, Either EvalError o)], st)
-evalBlockchain eval st (toNewestFirst . blocks -> blks) =
-    evalTraverse eval (concatMap (toList . blockData) (reverse blks)) st
+    -> (st, [Receipt c tx o])
```

Comment:
It is a bit obscure that any `EvalError`s are buried in the `Receipt`s 🤔

Reply:
Indeed - maybe making this more explicit by having it be a composition of two functions, like `intoReceipt . eval`

Reply:
We don’t return information about evaluation errors explicitly because blocks are valid even if a tx produces an eval error. This is why the `receiptTxOutput` field of a receipt has type `Either EvalError output` and not just `output`. If a transaction in the block produces an error we leave the state untouched and include the error in the receipt. But the transaction is a “valid” part of the blockchain.

However there’s a good case to be made that blocks should only contain valid transactions which would mean that `receiptTxOutput` just get’s type `output` and `evalBlock` would return `Either EvalError ...`. @cloudhead and I discussed this briefly and we are leaning towards this variant. Making this change requires a concerted effort though and should be done in another PR.

@cloudhead: Could you elaborate on your suggestion? I don’t get what you mean.

Reply:
I just mean that the `EvalError` is "swallowed" in the current implementation and not visible in the types, so having a more explicit `Either EvalError Output -> Receipt` step would surface the mechanism and be self-documenting.

Reply:
I'm not sure: is an empty block actually valid? Also, what happens to transactions which fail to evaluate, are they discarded?

Intuitively, I'd expect a signature like this:

```haskell
evalBlock 
    :: Hashable c tx
    => Evaluator st tx o
    -> st
    -> Block c tx s
    -> (st, Validation (NonEmpty EvalError) ([(EvalError, tx)], NonEmpty (Receipt c tx o))
```

ie. either a non-empty list of `EvalError`s if none of the transactions could be evaluated, or a pair of a possibly empty list of `EvalError`s along with the transactions which failed to evaluate, and a non-empty list of valid receipts. This type could be wrapped up in a `EvalResult` sum.

Reply:
> I'm not sure: is an empty block actually valid?

Depends on what you consider as valid. An empty block is accepted by `evalBlock`. Nakamoto consensus also needs empty blocks to work if the network does not produce transactions.

> I just mean that the `EvalError` is "swallowed" in the current implementation and not visible in the types, so having a more explicit `Either EvalError Output -> Receipt` step would surface the mechanism and be self-documenting.

This is just `evalTraverse` which is used by `evalBlock`.

To clarify I’m going to give an example. Assume a block has three transactions _tx₁_, _tx₂_, _tx₃_. Applying _tx₁_ to the start state _s₀_ produces _s₁_,  _tx₂_, fails and produces an error and _tx₃_ applied to  _s₁_ gives _s₂_. Then the block evaluates to the state _s₂_ and produces three receipts. The receipt for _tx₁_ includes the output of that transaction, the receipt for _tx₂_ includes an error and the receipt for _tx₃_ again holds the output of the transaction. All transactions are valid parts of the block and can be evaluated.

Reply:
I updated the tests to make this clearer

Reply:
> Nakamoto consensus also needs empty blocks to work if the network does not produce transactions.

Fair enough. I guess partitioning the receipts list to find erroneous transactions is fine then.